src/GafferScene/SceneWriter.cpp

```diff
@@ -283,8 +234,24 @@ void SceneWriter::executeSequence( const std::vector<float> &frames ) const
 			useSetsAPI = SceneReader::useSetsAPI( output.get() );
 		}
 
-		LocationWriter locationWriter( output.get(), !useSetsAPI ? sets.get() : nullptr, scope.context()->getTime(), mutex );
-		SceneAlgo::parallelProcessLocations( scene, locationWriter );
+		SceneAlgo::parallelGatherLocations(
+
+			scene,
+
+			// Collect LocationData from each location in parallel.
+
+			[&] ( const ScenePlug *scene, const ScenePlug::ScenePath &path ) {
+				return LocationData( scene, path, useSetsAPI ? nullptr : sets.get() );
+			},
+
+			// Write to output serially, because SceneInterfaces are not
+			// thread-safe for writing.
+
+			[&] ( const LocationData &locationData ) {
+				locationData.write( output.get(), scope.context()->getTime() );
+			}
+
+		);
```

Comment:
I feel like I'm grasping at straws a little trying to explain how this is failing. I did think it could be interesting that LocationData is getting an implicit assignment operator called - it's being directly stored in the queue, and it seems like there's an assignment happening in queue.pop()? It might be a performance concern that a bunch of smart pointers are getting pointlessly incremented/decremented, but I don't see any way it would affect correctness if LocationData does get copied, as long as write() gets called exactly once. ( Though I guess I don't know what a NameList is, and whether that is safe to copy ).

Reply:
I think `pop()` is using move assignment, and I've just pushed a commit to use that for `push()` too. So there shouldn't be any pointless increment/decrement now. NameList is just `vector<InternedString`.